### PR TITLE
Fix startAfter/endBefore for orderByKeys queries

### PR DESCRIPTION
**Pull request by IanWyszynski**
We should not be passing a fallback key to startAfter if we're querying on a key index. The indexValue is the key in this case.

### CODE REVIEW

File: firebase-database/src/main/java/com/google/firebase/database/Query.java [hunk 1]
```diff
@@ -331,7 +334,13 @@ public Query startAfter(boolean value) {
...
   public Query startAfter(@Nullable String value, @Nullable String key) {
     Node node =
-        value != null ? new StringNode(value, PriorityUtilities.NullPriority()) : EmptyNode.Empty();
+        value != null
+            ? new StringNode(
+                (params.getIndex().equals(KeyIndex.getInstance()))
+                    ? PushIdGenerator.successor(value)
+                    : value,
+                PriorityUtilities.NullPriority())
+            : EmptyNode.Empty();
```
Comment (schmidt-sebastian) [inline firebase-database/src/main/java/com/google/firebase/database/Query.java:343]: Can we rewrite this to match the style above? It's a bit hard to parse with two nested conditionals.

File: firebase-database/src/main/java/com/google/firebase/database/Query.java [hunk 2]
```diff
@@ -530,7 +542,13 @@ public Query endBefore(boolean value) {
...
   public Query endBefore(@Nullable String value, @Nullable String key) {
     Node node =
-        value != null ? new StringNode(value, PriorityUtilities.NullPriority()) : EmptyNode.Empty();
+        value != null
+            ? new StringNode(
+                params.getIndex().equals(KeyIndex.getInstance())
+                    ? PushIdGenerator.predecessor(value)
+                    : value,
+                PriorityUtilities.NullPriority())
+            : EmptyNode.Empty();
```
Comment (schmidt-sebastian) [inline firebase-database/src/main/java/com/google/firebase/database/Query.java:551]: Same here

**Review by schmidt-sebastian — APPROVED** [review]

File: firebase-database/src/main/java/com/google/firebase/database/Query.java (hunk 1 above)
Comment (schmidt-sebastian) [reply firebase-database/src/main/java/com/google/firebase/database/Query.java:343]: Also - do we not need the validation from https://github.com/firebase/firebase-ios-sdk/pull/7403/files#diff-fa09d2f832ea9e4ec0b8ca482585897a8706c7b6c2a0218250e08ed78c7c3c9aR205 here?

Comment (IanWyszynski) [reply firebase-database/src/main/java/com/google/firebase/database/Query.java:343]: I commented on that PR, but this case is covered in `validateQueryEndpoints`: 

https://github.com/firebase/firebase-android-sdk/blob/master/firebase-database/src/main/java/com/google/firebase/database/Query.java#L84-L103

File: firebase-database/src/main/java/com/google/firebase/database/Query.java [hunk 3]
```diff
@@ -115,18 +116,20 @@ private void validateQueryEndpoints(QueryParams params) {
...
               + "Use limitToFirst() or limitToLast() instead");
     }
   }
 
   /** This method validates that the equalTo call can be made */
   private void validateEqualToCall() {
     if (params.hasStart()) {
-      throw new IllegalArgumentException("Can't call equalTo() and startAt() combined");
+      throw new IllegalArgumentException(
+          "Can't call equalTo() and startAt() or startAfter() " + "combined");
```
Comment (schmidt-sebastian) [inline firebase-database/src/main/java/com/google/firebase/database/Query.java:128]:
```suggestion
          "Cannot combine equalTo() with startAt() or startAfter()");
```

File: firebase-database/src/main/java/com/google/firebase/database/Query.java [hunk 4]
```diff
@@ -115,18 +116,20 @@ private void validateQueryEndpoints(QueryParams params) {
...
     }
     if (params.hasEnd()) {
-      throw new IllegalArgumentException("Can't call equalTo() and endAt() combined");
+      throw new IllegalArgumentException(
+          "Can't call equalTo() and endAt() or startAfter() " + "combined");
```
Comment (schmidt-sebastian) [inline firebase-database/src/main/java/com/google/firebase/database/Query.java:132]:
```suggestion
          "Cannot combine equalTo() with endAt() or endBefore()");
```

**Review by schmidt-sebastian — APPROVED** [review]
Please update changelog.

File: firebase-database/src/androidTest/java/com/google/firebase/database/QueryTest.java [hunk 5]
```diff
@@ -581,6 +582,102 @@ public void setVariousLimitsWithStartAtName() throws DatabaseException, Interrup
...
+  public void testEndBeforeWithOrderByKeyOverlappingListener()
+      throws DatabaseException, InterruptedException, ExecutionException {
+    DatabaseReference ref = IntegrationTestHelpers.getRandomNode();
+    DatabaseReference childOne = ref.push();
+    DatabaseReference childTwo = ref.push();
+    Tasks.await(childOne.setValue(1L));
+    Tasks.await(childTwo.setValue(2L));
+
+    Semaphore semaphore = new Semaphore(0);
+    ref.addValueEventListener(
```
Comment (schmidt-sebastian) [inline firebase-database/src/androidTest/java/com/google/firebase/database/QueryTest.java:629]: We should remove this listener as well.

File: firebase-database/src/androidTest/java/com/google/firebase/database/QueryTest.java [hunk 6]
```diff
@@ -581,6 +582,102 @@ public void setVariousLimitsWithStartAtName() throws DatabaseException, Interrup
...
+    DataSnapshot snapshot = Tasks.await(ref.orderByKey().endBefore(childTwo.getKey()).get());
+    Map<String, Long> values = (Map<String, Long>) snapshot.getValue();
+
+    assertNotNull(values);
+    assertArrayEquals(values.keySet().toArray(), new String[] {childOne.getKey()});
+    assertArrayEquals(values.values().toArray(), new Long[] {values.get(childOne.getKey())});
+  }
+
+  @Test
+  public void testEndBeforeWithOrderByKeyOverlappingListener()
```
Comment (schmidt-sebastian) [inline firebase-database/src/androidTest/java/com/google/firebase/database/QueryTest.java:620]: Optional: Add a small description of the bug that this test addresses.

File: firebase-database/src/main/java/com/google/firebase/database/core/Repo.java [hunk 7]
```diff
@@ -496,12 +496,10 @@ public void onRequestResult(String optErrorCode, String optErrorMessage) {
           public void run() {
             // Always check active-listener in-memory caches first. These are always at least as
             // up to date as the persistence cache.
-            Node cached =
-                serverSyncTree.calcCompleteEventCacheFromRoot(query.getPath(), new ArrayList<>());
-            if (!cached.isEmpty()) {
+            Node cached = serverSyncTree.getServerValue(query.getSpec());
```
Comment (schmidt-sebastian) [inline firebase-database/src/main/java/com/google/firebase/database/core/Repo.java:499]: s/cached/serverValue ?

File: firebase-database/src/main/java/com/google/firebase/database/core/SyncPoint.java [hunk 8]
```diff
@@ -113,11 +113,8 @@ public boolean isEmpty() {
   }
 
   /** Add an event callback for the specified query. */
-  public List<DataEvent> addEventRegistration(
-      @NotNull EventRegistration eventRegistration,
-      WriteTreeRef writesCache,
-      CacheNode serverCache) {
-    QuerySpec query = eventRegistration.getQuerySpec();
+  public View getView(QuerySpec query, WriteTreeRef writesCache, CacheNode serverCache) {
+    // TODO: Inline with addEventRegistration
```
Comment (schmidt-sebastian) [inline firebase-database/src/main/java/com/google/firebase/database/core/SyncPoint.java:117]: This TODO has been addressed, hasn't it?

File: firebase-database/src/main/java/com/google/firebase/database/core/SyncTree.java [hunk 9]
```diff
@@ -484,6 +484,63 @@ public DataSnapshot persistenceServerCache(Query query) {
...
+    return persistenceManager.runInTransaction(
+        () -> {
+          Path path = query.getPath();
+
+          Node serverCacheNode = null;
+          boolean foundAncestorDefaultView = false;
+          // Any covering writes will necessarily be at the root, so really all we need to find is
+          // the server cache. Consider optimizing this once there's a better understanding of
+          // what actual behavior will be.
+          // for (Map.Entry<QuerySpec, View> entry: views.entrySet()) {
```
Comment (schmidt-sebastian) [inline firebase-database/src/main/java/com/google/firebase/database/core/SyncTree.java:498]: Please fix the commented out code here (by removing I assume)

File: firebase-database/src/main/java/com/google/firebase/database/core/SyncTree.java [hunk 10]
```diff
@@ -484,6 +484,63 @@ public DataSnapshot persistenceServerCache(Query query) {
...
+          {
```
Comment (schmidt-sebastian) [inline firebase-database/src/main/java/com/google/firebase/database/core/SyncTree.java:499]: Do we need these braces here?

**Review by schmidt-sebastian — APPROVED** [review]

**Review by schmidt-sebastian — APPROVED** [review]
Still LGTM

File: firebase-database/CHANGELOG.md [hunk 11]
```diff
@@ -1,3 +1,6 @@
+# 19.7.1
```
Comment (schmidt-sebastian) [inline firebase-database/CHANGELOG.md:1]: I just realized that we can probably drop this since we never released startAfter(). I can do this manually when we work on the public release notes if you don't want to do another CI run.
